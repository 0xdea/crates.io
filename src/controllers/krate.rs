use axum::extract::{FromRequestParts, Path};

pub mod delete;
pub mod downloads;
pub mod follow;
pub mod metadata;
pub mod owners;
pub mod publish;
pub mod search;
pub mod versions;

#[derive(Deserialize, FromRequestParts)]
#[from_request(via(Path))]
pub struct CratePath {
    /// Name of the crate
    pub name: String,
}
