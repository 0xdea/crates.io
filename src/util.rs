pub use self::bytes_request::BytesRequest;
pub use self::io_util::{read_fill, read_le_u32};
pub use self::request_helpers::*;

mod bytes_request;
pub mod diesel;
pub mod errors;
mod io_util;
mod request_helpers;
pub mod rfc3339;
pub mod token;
pub mod tracing;
