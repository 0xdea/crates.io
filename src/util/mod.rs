use std::io::{MemReader, IoError};
use std::io::process::{ProcessOutput, Command};
use std::collections::HashMap;
use std::fmt::Show;
use std::str;

use serialize::{json, Encodable};
use url;

use conduit::{Request, Response, Handler};

pub use self::errors::{CargoError, CargoResult, internal, human, internal_error};
pub use self::errors::{ChainError, BoxError};
pub use self::result::{Require, Wrap};
pub use self::lazy_cell::LazyCell;
pub use self::io::LimitErrorReader;
pub use self::hasher::HashingReader;

pub mod errors;
pub mod result;
mod lazy_cell;
mod io;
mod hasher;

pub trait RequestUtils {
    fn redirect(self, url: String) -> Response;

    fn json<'a, T: Encodable<json::Encoder<'a>, IoError>>(self, t: &T) -> Response;
    fn query(self) -> HashMap<String, String>;
    fn wants_json(self) -> bool;
}

pub fn json_response<'a, T>(t: &T) -> Response
                            where T: Encodable<json::Encoder<'a>, IoError> {
    let s = json::encode(t);
    let mut headers = HashMap::new();
    headers.insert("Content-Type".to_string(),
                   vec!["application/json; charset=utf-8".to_string()]);
    Response {
        status: (200, "OK"),
        headers: headers,
        body: box MemReader::new(s.into_bytes()),
    }
}


impl<'a> RequestUtils for &'a Request + 'a {
    fn json<'a, T: Encodable<json::Encoder<'a>, IoError>>(self, t: &T) -> Response {
        json_response(t)
    }

    fn query(self) -> HashMap<String, String> {
        url::form_urlencoded::parse_str(self.query_string().unwrap_or(""))
            .into_iter().collect()
    }

    fn redirect(self, url: String) -> Response {
        let mut headers = HashMap::new();
        headers.insert("Location".to_string(), vec![url.to_string()]);
        Response {
            status: (302, "Found"),
            headers: headers,
            body: box MemReader::new(Vec::new()),
        }
    }

    fn wants_json(self) -> bool {
        let content = self.headers().find("Accept").unwrap_or(Vec::new());
        content.iter().any(|s| s.contains("json"))
    }
}

pub struct C(pub fn(&mut Request) -> CargoResult<Response>);

impl Handler for C {
    fn call(&self, req: &mut Request) -> Result<Response, Box<Show + 'static>> {
        let C(f) = *self;
        match f(req) {
            Ok(req) => Ok(req),
            Err(e) => {
                match e.response() {
                    Some(response) => Ok(response),
                    None => Err(box e as Box<Show>),
                }
            }
        }
    }
}

pub fn exec(cmd: &Command) -> CargoResult<ProcessOutput> {
    let output = try!(cmd.output().chain_error(|| {
        internal(format!("failed to run command `{}`", cmd))
    }));
    if !output.status.success() {
        let mut desc = String::new();
        if output.output.len() != 0 {
            desc.push_str("--- stdout\n");
            desc.push_str(str::from_utf8(output.output.as_slice()).unwrap());
        }
        if output.error.len() != 0 {
            desc.push_str("--- stderr\n");
            desc.push_str(str::from_utf8(output.error.as_slice()).unwrap());
        }
        Err(internal_error(format!("failed to run command `{}`", cmd), desc))
    } else {
        Ok(output)
    }
}
