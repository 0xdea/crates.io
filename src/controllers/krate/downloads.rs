//! Endpoint for exposing crate download counts
//!
//! The endpoint for downloading a crate and exposing version specific
//! download counts are located in `version::downloads`.

use crate::app::AppState;
use crate::controllers::krate::CratePath;
use crate::models::download::Version;
use crate::models::{User, Version as FullVersion, VersionDownload, VersionOwnerAction};
use crate::schema::{version_downloads, version_owner_actions, versions};
use crate::util::errors::{bad_request, AppResult, BoxedAppError};
use crate::views::{EncodableVersion, EncodableVersionDownload};
use axum::extract::FromRequestParts;
use axum_extra::extract::Query;
use axum_extra::json;
use axum_extra::response::ErasedJson;
use crates_io_database::schema::users;
use crates_io_diesel_helpers::to_char;
use diesel::prelude::*;
use diesel_async::{AsyncPgConnection, RunQueryDsl};
use futures_util::future::BoxFuture;
use futures_util::FutureExt;
use std::cmp;
use std::str::FromStr;

#[derive(Debug, Deserialize, FromRequestParts, utoipa::IntoParams)]
#[from_request(via(Query))]
#[into_params(parameter_in = Query)]
pub struct DownloadsQueryParams {
    /// Additional data to include in the response.
    ///
    /// Valid values: `versions`.
    ///
    /// Defaults to no additional data.
    ///
    /// This parameter expects a comma-separated list of values.
    include: Option<String>,
}

/// Get the download counts for a crate.
///
/// This includes the per-day downloads for the last 90 days and for the
/// latest 5 versions plus the sum of the rest.
#[utoipa::path(
    get,
    path = "/api/v1/crates/{name}/downloads",
    params(CratePath, DownloadsQueryParams),
    tag = "crates",
    responses((status = 200, description = "Successful Response")),
)]

pub async fn get_crate_downloads(
    state: AppState,
    path: CratePath,
    params: DownloadsQueryParams,
) -> AppResult<ErasedJson> {
    let mut conn = state.db_read().await?;

    use diesel::dsl::*;
    use diesel::sql_types::BigInt;

    let crate_id: i32 = path.load_crate_id(&mut conn).await?;

    let mut versions: Vec<Version> = versions::table
        .filter(versions::crate_id.eq(crate_id))
        .select(Version::as_select())
        .load(&mut conn)
        .await?;

    versions.sort_unstable_by(|a, b| b.num.cmp(&a.num));
    let (latest_five, rest) = versions.split_at(cmp::min(5, versions.len()));

    let include = params
        .include
        .as_ref()
        .map(|mode| ShowIncludeMode::from_str(mode))
        .transpose()?
        .unwrap_or_default();

    let sum_downloads = sql::<BigInt>("SUM(version_downloads.downloads)");
    let (downloads, extra, versions_and_publishers, actions) = tokio::try_join!(
        VersionDownload::belonging_to(latest_five)
            .filter(version_downloads::date.gt(date(now - 90.days())))
            .order((
                version_downloads::date.asc(),
                version_downloads::version_id.desc(),
            ))
            .load(&mut conn)
            .boxed(),
        VersionDownload::belonging_to(rest)
            .select((
                to_char(version_downloads::date, "YYYY-MM-DD"),
                sum_downloads,
            ))
            .filter(version_downloads::date.gt(date(now - 90.days())))
            .group_by(version_downloads::date)
            .order(version_downloads::date.asc())
            .load::<ExtraDownload>(&mut conn)
            .boxed(),
        load_versions_and_publishers(&mut conn, latest_five, include.versions),
        load_actions(&mut conn, latest_five, include.versions),
    )?;

    let downloads = downloads
        .into_iter()
        .map(VersionDownload::into)
        .collect::<Vec<EncodableVersionDownload>>();

    #[derive(Serialize, Queryable)]
    struct ExtraDownload {
        date: String,
        downloads: i64,
    }

    if include.versions {
        let versions_and_publishers = versions_and_publishers.grouped_by(latest_five);
        let actions = actions.grouped_by(latest_five);
        let versions = versions_and_publishers
            .into_iter()
            .zip(actions)
            .filter_map(|(vp, actions)| {
                vp.into_iter().next().map(|(version, publisher)| {
                    EncodableVersion::from(version, &path.name, publisher, actions)
                })
            })
            .collect::<Vec<_>>();

        return Ok(json!({
            "version_downloads": downloads,
            "versions": versions,
            "meta": {
                "extra_downloads": extra,
            },
        }));
    }

    Ok(json!({
        "version_downloads": downloads,
        "meta": {
            "extra_downloads": extra,
        },
    }))
}

type VersionsAndPublishers = (FullVersion, Option<User>);
fn load_versions_and_publishers<'a>(
    conn: &mut AsyncPgConnection,
    versions: &'a [Version],
    includes: bool,
) -> BoxFuture<'a, QueryResult<Vec<VersionsAndPublishers>>> {
    if !includes {
        return futures_util::future::always_ready(|| Ok(vec![])).boxed();
    }
    FullVersion::belonging_to(versions)
        .left_outer_join(users::table)
        .select(VersionsAndPublishers::as_select())
        .load(conn)
        .boxed()
}

fn load_actions<'a>(
    conn: &mut AsyncPgConnection,
    versions: &'a [Version],
    includes: bool,
) -> BoxFuture<'a, QueryResult<Vec<(VersionOwnerAction, User)>>> {
    if !includes {
        return futures_util::future::always_ready(|| Ok(vec![])).boxed();
    }
    VersionOwnerAction::belonging_to(versions)
        .inner_join(users::table)
        .order(version_owner_actions::id)
        .load(conn)
        .boxed()
}

#[derive(Debug, Default)]
struct ShowIncludeMode {
    versions: bool,
}

impl ShowIncludeMode {
    const INVALID_COMPONENT: &'static str = "invalid component for ?include= (expected 'versions')";
}

impl FromStr for ShowIncludeMode {
    type Err = BoxedAppError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let mut mode = Self { versions: false };
        for component in s.split(',') {
            match component {
                "" => {}
                "versions" => mode.versions = true,
                _ => return Err(bad_request(Self::INVALID_COMPONENT)),
            }
        }
        Ok(mode)
    }
}
