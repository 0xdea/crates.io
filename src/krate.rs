use std::collections::HashMap;
use std::sync::Arc;
use std::time::Duration;
use serialize::hex::ToHex;
use serialize::json;
use time::Timespec;

use conduit::{Request, Response};
use conduit_router::RequestParams;
use curl::http;
use pg::types::ToSql;
use pg::{PostgresRow, PostgresStatement};
use semver;

use app::{App, RequestApp};
use db::{Connection, RequestTransaction};
use download::{VersionDownload, EncodableVersionDownload};
use git;
use model::Model;
use upload;
use user::{User, RequestUser};
use util::errors::{NotFound, CargoError};
use util::{LimitErrorReader, HashingReader};
use util::{RequestUtils, CargoResult, Require, internal, ChainError, human};
use version::{Version, EncodableVersion};

#[deriving(Clone)]
pub struct Crate {
    pub id: i32,
    pub name: String,
    pub user_id: i32,
    pub updated_at: Timespec,
    pub created_at: Timespec,
    pub downloads: i32,
    pub max_version: semver::Version,
}

#[deriving(Encodable, Decodable)]
pub struct EncodableCrate {
    pub id: String,
    pub name: String,
    pub updated_at: String,
    pub versions: Option<Vec<i32>>,
    pub created_at: String,
    pub downloads: i32,
    pub max_version: String,
    pub links: CrateLinks,
}

#[deriving(Encodable, Decodable)]
pub struct CrateLinks {
    pub version_downloads: String,
    pub versions: Option<String>,
}

impl Crate {
    pub fn find(conn: &Connection, id: i32) -> CargoResult<Crate> {
        Model::find(conn, id)
    }

    pub fn find_by_name(conn: &Connection, name: &str) -> CargoResult<Crate> {
        let stmt = try!(conn.prepare("SELECT * FROM crates \
                                      WHERE name = $1 LIMIT 1"));
        match try!(stmt.query(&[&name as &ToSql])).next() {
            Some(row) => Ok(Model::from_row(&row)),
            None => Err(NotFound.box_error()),
        }
    }

    pub fn find_or_insert(conn: &Connection, name: &str,
                          user_id: i32) -> CargoResult<Crate> {
        // TODO: like with users, this is sadly racy

        let stmt = try!(conn.prepare("SELECT * FROM crates WHERE name = $1"));
        let mut rows = try!(stmt.query(&[&name as &ToSql]));
        match rows.next() {
            Some(row) => return Ok(Model::from_row(&row)),
            None => {}
        }
        let stmt = try!(conn.prepare("INSERT INTO crates \
                                      (name, user_id, created_at,
                                       updated_at, downloads, max_version) \
                                      VALUES ($1, $2, $3, $4, 0, '0.0.0') \
                                      RETURNING *"));
        let now = ::now();
        let mut rows = try!(stmt.query(&[&name as &ToSql, &user_id, &now, &now]));
        Ok(Model::from_row(&try!(rows.next().require(|| {
            internal("no crate returned")
        }))))
    }

    pub fn valid_name(name: &str) -> bool {
        if name.len() == 0 { return false }
        name.chars().all(|c| c.is_alphanumeric() || c == '_' || c == '-')
    }

    pub fn encodable(self, versions: Option<Vec<i32>>) -> EncodableCrate {
        let Crate { name, created_at, updated_at, downloads,
                    max_version, .. } = self;
        let versions_link = match versions {
            Some(..) => None,
            None => Some(format!("/crates/{}/versions", name)),
        };
        EncodableCrate {
            id: name.clone(),
            name: name.clone(),
            updated_at: ::encode_time(updated_at),
            created_at: ::encode_time(created_at),
            downloads: downloads,
            versions: versions,
            max_version: max_version.to_string(),
            links: CrateLinks {
                version_downloads: format!("/crates/{}/downloads", name),
                versions: versions_link,
            },
        }
    }

    pub fn versions(&self, conn: &Connection) -> CargoResult<Vec<Version>> {
        let stmt = try!(conn.prepare("SELECT * FROM versions \
                                      WHERE crate_id = $1"));
        let rows = try!(stmt.query(&[&self.id]));
        Ok(rows.map(|r| Model::from_row(&r)).collect())
    }

    pub fn s3_path(&self, version: &str) -> String {
        format!("/pkg/{}/{}-{}.tar.gz", self.name, self.name, version)
    }

    pub fn add_version(&mut self, conn: &Connection, ver: &semver::Version,
                       features: &HashMap<String, Vec<String>>)
                       -> CargoResult<Version> {
        match try!(Version::find_by_num(conn, self.id, ver)) {
            Some(..) => {
                return Err(human(format!("crate version `{}` is already uploaded",
                                         ver)))
            }
            None => {}
        }
        if *ver > self.max_version { self.max_version = ver.clone(); }
        self.updated_at = ::now();
        try!(conn.execute("UPDATE crates SET updated_at = $1, max_version = $2
                           WHERE id = $3",
                          &[&self.updated_at, &self.max_version.to_string(),
                            &self.id]));
        Version::insert(conn, self.id, ver, features)
    }
}

impl Model for Crate {
    fn from_row(row: &PostgresRow) -> Crate {
        let max: String = row.get("max_version");
        Crate {
            id: row.get("id"),
            name: row.get("name"),
            user_id: row.get("user_id"),
            updated_at: row.get("updated_at"),
            created_at: row.get("created_at"),
            downloads: row.get("downloads"),
            max_version: semver::Version::parse(max.as_slice()).unwrap(),
        }
    }
    fn table_name(_: Option<Crate>) -> &'static str { "crates" }
}

pub fn index(req: &mut Request) -> CargoResult<Response> {
    let conn = try!(req.tx());
    let (offset, limit) = try!(req.pagination(10, 100));
    let query = req.query();
    let sort = query.find_equiv(&"sort").map(|s| s.as_slice()).unwrap_or("alpha");
    let sort_sql = match sort {
        "downloads" => "ORDER BY crates.downloads DESC",
        _ => "ORDER BY crates.name ASC",
    };

    // Different queries for different parameters
    let mut pattern;
    let mut id;
    let mut args = vec![&limit as &ToSql, &offset as &ToSql];
    let (q, cnt) = match (query.find_equiv(&"q"), query.find_equiv(&"letter")) {
        (Some(query), _) => {
            args.insert(0, query as &ToSql);
            ("SELECT crates.* FROM crates,
                                   plainto_tsquery($1) q,
                                   to_tsvector('english', name) txt,
                                   ts_rank_cd(txt, q) rank
              WHERE q @@ txt
              ORDER BY rank DESC LIMIT $2 OFFSET $3".to_string(),
             "SELECT COUNT(crates.*) FROM crates,
                                          plainto_tsquery($1) q,
                                          to_tsvector('english', name) txt
              WHERE q @@ txt")
        }
        (None, Some(letter)) => {
            pattern = format!("{}%", letter.as_slice().char_at(0)
                                           .to_lowercase());
            args.insert(0, &pattern as &ToSql);
            (format!("SELECT * FROM crates WHERE name LIKE $1 {}
                      LIMIT $2 OFFSET $3", sort_sql),
             "SELECT COUNT(*) FROM crates WHERE name LIKE $1")
        },
        (None, None) => {
            let user_id = query.find_equiv(&"user_id").map(|s| s.as_slice())
                               .and_then(from_str::<i32>);
            let following = query.find_equiv(&"following").is_some();
            match (user_id, following) {
                (Some(user_id), _) => {
                    id = user_id;
                    args.insert(0, &id as &ToSql);
                    (format!("SELECT * FROM crates WHERE user_id = $1 {} \
                              LIMIT $2 OFFSET $3",
                             sort_sql),
                     "SELECT COUNT(*) FROM crates WHERE user_id = $1")
                }
                (None, true) => {
                    let me = try!(req.user());
                    id = me.id;
                    args.insert(0, &id as &ToSql);
                    (format!("SELECT crates.* FROM crates
                              INNER JOIN follows
                                 ON follows.crate_id = crates.id AND
                                    follows.user_id = $1
                              {} LIMIT $2 OFFSET $3", sort_sql),
                     "SELECT COUNT(crates.*) FROM crates
                      INNER JOIN follows
                         ON follows.crate_id = crates.id AND
                            follows.user_id = $1")
                }
                (None, false) => {
                    (format!("SELECT * FROM crates {} LIMIT $1 OFFSET $2",
                             sort_sql),
                     "SELECT COUNT(*) FROM crates")
                }
            }
        }
    };

    // Collect all the crates
    let stmt = try!(conn.prepare(q.as_slice()));
    let mut crates = Vec::new();
    for row in try!(stmt.query(args.as_slice())) {
        let krate: Crate = Model::from_row(&row);
        crates.push(krate.encodable(None));
    }

    // Query for the total count of crates
    let stmt = try!(conn.prepare(cnt));
    let args = if args.len() > 2 {args.slice_to(1)} else {args.slice_to(0)};
    let row = try!(stmt.query(args)).next().unwrap();
    let total = row.get(0u);

    #[deriving(Encodable)]
    struct R { crates: Vec<EncodableCrate>, meta: Meta }
    #[deriving(Encodable)]
    struct Meta { total: i64 }

    Ok(req.json(&R {
        crates: crates,
        meta: Meta { total: total },
    }))
}

pub fn summary(req: &mut Request) -> CargoResult<Response> {
    let tx = try!(req.tx());
    let num_crates = {
        let stmt = try!(tx.prepare("SELECT COUNT(*) FROM crates"));
        let mut rows = try!(stmt.query(&[]));
        rows.next().unwrap().get("count")
    };
    let num_downloads = {
        let stmt = try!(tx.prepare("SELECT total_downloads FROM metadata"));
        let mut rows = try!(stmt.query(&[]));
        rows.next().unwrap().get("total_downloads")
    };

    let to_crates = |stmt: PostgresStatement| {
        let rows = raw_try!(stmt.query([]));
        Ok(rows.map(|r| {
            let krate: Crate = Model::from_row(&r);
            krate.encodable(None)
        }).collect::<Vec<EncodableCrate>>())
    };
    let new_crates = try!(tx.prepare("SELECT * FROM crates \
                                        ORDER BY created_at DESC LIMIT 10"));
    let just_updated = try!(tx.prepare("SELECT * FROM crates \
                                        ORDER BY updated_at DESC LIMIT 10"));
    let most_downloaded = try!(tx.prepare("SELECT * FROM crates \
                                           ORDER BY downloads DESC LIMIT 10"));

    #[deriving(Encodable)]
    struct R {
        num_downloads: i64,
        num_crates: i64,
        new_crates: Vec<EncodableCrate>,
        most_downloaded: Vec<EncodableCrate>,
        just_updated: Vec<EncodableCrate>,
    }
    Ok(req.json(&R {
        num_downloads: num_downloads,
        num_crates: num_crates,
        new_crates: try!(to_crates(new_crates)),
        most_downloaded: try!(to_crates(most_downloaded)),
        just_updated: try!(to_crates(just_updated)),
    }))
}

pub fn show(req: &mut Request) -> CargoResult<Response> {
    let name = &req.params()["crate_id"];
    let conn = try!(req.tx());
    let krate = try!(Crate::find_by_name(&*conn, name.as_slice()));
    let versions = try!(krate.versions(&*conn));
    let ids = versions.iter().map(|v| v.id).collect();

    #[deriving(Encodable)]
    struct R { krate: EncodableCrate, versions: Vec<EncodableVersion>, }
    Ok(req.json(&R {
        krate: krate.clone().encodable(Some(ids)),
        versions: versions.into_iter().map(|v| {
            v.encodable(krate.name.as_slice())
        }).collect(),
    }))
}

pub fn new(req: &mut Request) -> CargoResult<Response> {
    let app = req.app().clone();

    let (new_crate, user) = try!(parse_new_headers(req));
    let name = new_crate.name.as_slice();
    let vers = &*new_crate.vers;
    let features = new_crate.features.iter().map(|(k, v)| {
        ((**k).to_string(), v.iter().map(|v| (**v).to_string()).collect())
    }).collect::<HashMap<String, Vec<String>>>();

    // Persist the new crate, if it doesn't already exist
    let mut krate = try!(Crate::find_or_insert(try!(req.tx()), name, user.id));
    if krate.user_id != user.id {
        return Err(human("crate name has already been claimed by another user"))
    }

    // Persist the new version of this crate
    let mut version = try!(krate.add_version(try!(req.tx()), vers, &features));

    // Link this new version to all dependencies
    let mut deps = Vec::new();
    for dep in new_crate.deps.iter() {
        let (dep, krate) = try!(version.add_dependency(try!(req.tx()), dep));
        deps.push(dep.git_encode(krate.name.as_slice()));
    }

    // Upload the crate to S3
    let handle = http::handle();
    let mut handle = match req.app().s3_proxy {
        Some(ref proxy) => handle.proxy(proxy.as_slice()),
        None => handle,
    };
    let path = krate.s3_path(vers.to_string().as_slice());
    let (resp, cksum) = {
        let length = try!(req.body().read_le_u32());
        let body = LimitErrorReader::new(req.body(), app.config.max_upload_size);
        let mut body = HashingReader::new(body);
        let resp = {
            let s3req = app.bucket.put(&mut handle, path.as_slice(), &mut body,
                                       "application/x-tar")
                                  .content_length(length as uint)
                                  .header("Content-Encoding", "gzip");
            try!(s3req.exec().chain_error(|| {
                internal(format!("failed to upload to S3: `{}`", path))
            }))
        };
        (resp, body.final())
    };
    if resp.get_code() != 200 {
        return Err(internal(format!("failed to get a 200 response from S3: {}",
                                    resp)))
    }

    // If the git commands fail below, we shouldn't keep the crate on the
    // server.
    struct Bomb { app: Arc<App>, path: Option<String>, handle: http::Handle }
    impl Drop for Bomb {
        fn drop(&mut self) {
            match self.path {
                Some(ref path) => {
                    let _ = self.app.bucket.delete(&mut self.handle,
                                                   path.as_slice())
                                .exec();
                }
                None => {}
            }
        }
    }
    let mut bomb = Bomb { app: app.clone(), path: Some(path), handle: handle };

    // Register this crate in our local git repo.
    let git_crate = git::GitCrate {
        name: name.to_string(),
        vers: vers.to_string(),
        cksum: cksum.as_slice().to_hex(),
        features: features,
        deps: deps,
    };
    try!(git::add_crate(&**req.app(), &git_crate).chain_error(|| {
        internal(format!("could not add crate `{}` to the git repo", name))
    }));

    // Now that we've come this far, we're committed!
    bomb.path = None;

    #[deriving(Encodable)]
    struct R { krate: EncodableCrate }
    Ok(req.json(&R { krate: krate.encodable(None) }))
}

fn parse_new_headers(req: &mut Request) -> CargoResult<(upload::NewCrate, User)> {
    // Make sure the tarball being uploaded looks sane
    let length = try!(req.content_length().require(|| {
        human("missing header: Content-Length")
    }));
    let max = req.app().config.max_upload_size;
    if length > max { return Err(human(format!("max upload size is: {}", max))) }

    // Read the json upload request
    let amt = try!(req.body().read_le_u32()) as uint;
    if amt > max { return Err(human(format!("max upload size is: {}", max))) }
    let json = try!(req.body().read_exact(amt));
    let json = try!(String::from_utf8(json).map_err(|_| {
        human("json body was not valid utf-8")
    }));
    let new: upload::NewCrate = try!(json::decode(json.as_slice()).map_err(|e| {
        human(format!("invalid upload request: {}", e))
    }));

    // Peel out authentication
    fn header<'a>(req: &'a Request, hdr: &str) -> CargoResult<Vec<&'a str>> {
        req.headers().find(hdr).require(|| {
            human(format!("missing header: {}", hdr))
        })
    }
    let auth = try!(header(req, "X-Cargo-Auth"))[0].to_string();

    // Make sure the api token is a valid api token
    let user = try!(User::find_by_api_token(try!(req.tx()),
                                            auth.as_slice()).map_err(|_| {
        human("invalid or unknown auth token supplied")
    }));

    Ok((new, user))
}

pub fn download(req: &mut Request) -> CargoResult<Response> {
    let crate_name = req.params()["crate_id"].as_slice();
    let version = req.params()["version"].as_slice();
    let tx = try!(req.tx());
    let stmt = try!(tx.prepare("SELECT versions.id as version_id
                                FROM crates
                                LEFT JOIN versions ON
                                    crates.id = versions.crate_id
                                WHERE crates.name = $1
                                  AND versions.num = $2
                                LIMIT 1"));
    let mut rows = try!(stmt.query(&[&crate_name as &ToSql, &version as &ToSql]));
    let row = try!(rows.next().require(|| human("crate or version not found")));
    let version_id: i32 = row.get("version_id");
    let now = ::now();

    // Bump download counts.
    //
    // Note that this is *not* an atomic update, and that's somewhat
    // intentional. It doesn't appear that postgres supports an atomic update of
    // a counter, so we just do the hopefully "least racy" thing. This is
    // largely ok because these download counters are just that, counters. No
    // need to have super high-fidelity counter.
    //
    // Also, we only update the counter for *today*, nothing else. We have lots
    // of other counters, but they're all updated later on via the
    // update-downloads script.
    let amt = try!(tx.execute("UPDATE version_downloads
                               SET downloads = downloads + 1
                               WHERE version_id = $1 AND date($2) = date(date)",
                              &[&version_id, &now]));
    if amt == 0 {
        try!(tx.execute("INSERT INTO version_downloads
                         (version_id, downloads, counted, date, processed)
                         VALUES ($1, 1, 0, date($2), false)",
                        &[&version_id, &now]));
    }

    // Now that we've done our business, redirect to the actual data.
    let redirect_url = format!("https://{}/pkg/{}/{}-{}.tar.gz",
                               req.app().bucket.host(),
                               crate_name, crate_name, version);

    if req.wants_json() {
        #[deriving(Encodable)]
        struct R { url: String }
        Ok(req.json(&R{ url: redirect_url }))
    } else {
        Ok(req.redirect(redirect_url))
    }
}

pub fn downloads(req: &mut Request) -> CargoResult<Response> {
    let crate_name = req.params()["crate_id"].as_slice();
    let tx = try!(req.tx());
    let krate = try!(Crate::find_by_name(tx, crate_name));

    let cutoff_date = ::now() + Duration::days(-90);
    let stmt = try!(tx.prepare("SELECT * FROM version_downloads
                                LEFT JOIN versions
                                    ON versions.id = version_downloads.version_id
                                WHERE date > $1 AND versions.crate_id = $2
                                ORDER BY date ASC"));
    let mut downloads = Vec::new();
    for row in try!(stmt.query(&[&cutoff_date, &krate.id])) {
        let download: VersionDownload = Model::from_row(&row);
        downloads.push(download.encodable());
    }

    #[deriving(Encodable)]
    struct R { version_downloads: Vec<EncodableVersionDownload> }
    Ok(req.json(&R{ version_downloads: downloads }))
}

fn user_and_crate(req: &mut Request) -> CargoResult<(User, Crate)> {
    let user = try!(req.user());
    let crate_name = req.params()["crate_id"].as_slice();
    let tx = try!(req.tx());
    let krate = try!(Crate::find_by_name(tx, crate_name));
    Ok((user.clone(), krate))
}

pub fn follow(req: &mut Request) -> CargoResult<Response> {
    let (user, krate) = try!(user_and_crate(req));
    let tx = try!(req.tx());
    let stmt = try!(tx.prepare("SELECT 1 FROM follows
                                WHERE user_id = $1 AND crate_id = $2"));
    let mut rows = try!(stmt.query(&[&user.id, &krate.id]));
    if !rows.next().is_some() {
        try!(tx.execute("INSERT INTO follows (user_id, crate_id)
                         VALUES ($1, $2)", &[&user.id, &krate.id]));
    }
    #[deriving(Encodable)]
    struct R { ok: bool }
    Ok(req.json(&R { ok: true }))
}

pub fn unfollow(req: &mut Request) -> CargoResult<Response> {
    let (user, krate) = try!(user_and_crate(req));
    let tx = try!(req.tx());
    try!(tx.execute("DELETE FROM follows
                     WHERE user_id = $1 AND crate_id = $2",
                    &[&user.id, &krate.id]));
    #[deriving(Encodable)]
    struct R { ok: bool }
    Ok(req.json(&R { ok: true }))
}

pub fn following(req: &mut Request) -> CargoResult<Response> {
    let (user, krate) = try!(user_and_crate(req));
    let tx = try!(req.tx());
    let stmt = try!(tx.prepare("SELECT 1 FROM follows
                                WHERE user_id = $1 AND crate_id = $2"));
    let mut rows = try!(stmt.query(&[&user.id, &krate.id]));
    #[deriving(Encodable)]
    struct R { following: bool }
    Ok(req.json(&R { following: rows.next().is_some() }))
}

pub fn versions(req: &mut Request) -> CargoResult<Response> {
    let crate_name = req.params()["crate_id"].as_slice();
    let tx = try!(req.tx());
    let krate = try!(Crate::find_by_name(tx, crate_name));

    let stmt = try!(tx.prepare("SELECT * FROM versions WHERE crate_id = $1"));
    let mut versions = Vec::new();
    for row in try!(stmt.query(&[&krate.id])) {
        let version: Version = Model::from_row(&row);
        versions.push(version.encodable(crate_name));
    }

    #[deriving(Encodable)]
    struct R { versions: Vec<EncodableVersion> }
    Ok(req.json(&R{ versions: versions }))
}
