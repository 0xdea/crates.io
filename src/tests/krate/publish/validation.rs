use crate::builders::PublishBuilder;
use crate::util::{RequestHelper, TestApp};
use crates_io::models::krate::MAX_NAME_LENGTH;
use http::StatusCode;
use insta::assert_yaml_snapshot;

#[test]
fn invalid_names() {
    let (app, _, _, token) = TestApp::full().with_token();

    let bad_name = |name: &str, error_message: &str| {
        let crate_to_publish = PublishBuilder::new(name, "1.0.0");
        let response = token.publish_crate(crate_to_publish);
        assert_eq!(response.status(), StatusCode::OK);

        let json = response.into_json();
        let json = json.as_object().unwrap();
        let errors = json.get("errors").unwrap().as_array().unwrap();
        let first_error = errors.first().unwrap().as_object().unwrap();
        let detail = first_error.get("detail").unwrap().as_str().unwrap();
        assert!(detail.contains(error_message), "{detail:?}");
    };

    let error_message = "expected a valid crate name";
    bad_name("", error_message);
    bad_name("foo bar", error_message);
    bad_name(&"a".repeat(MAX_NAME_LENGTH + 1), error_message);
    bad_name("snow☃", error_message);
    bad_name("áccênts", error_message);

    let error_message = "cannot upload a crate with a reserved name";
    bad_name("std", error_message);
    bad_name("STD", error_message);
    bad_name("compiler-rt", error_message);
    bad_name("compiler_rt", error_message);
    bad_name("coMpiLer_Rt", error_message);

    assert!(app.stored_files().is_empty());
}

#[test]
fn license_and_description_required() {
    let (app, _, _, token) = TestApp::full().with_token();

    let crate_to_publish = PublishBuilder::new("foo_metadata", "1.1.0")
        .unset_license()
        .unset_description();

    let response = token.publish_crate(crate_to_publish);
    assert_eq!(response.status(), StatusCode::OK);
    assert_yaml_snapshot!(response.into_json());

    let crate_to_publish = PublishBuilder::new("foo_metadata", "1.1.0").unset_description();

    let response = token.publish_crate(crate_to_publish);
    assert_eq!(response.status(), StatusCode::OK);
    assert_yaml_snapshot!(response.into_json());

    let crate_to_publish = PublishBuilder::new("foo_metadata", "1.1.0")
        .unset_license()
        .license_file("foo")
        .unset_description();

    let response = token.publish_crate(crate_to_publish);
    assert_eq!(response.status(), StatusCode::OK);
    assert_yaml_snapshot!(response.into_json());

    assert!(app.stored_files().is_empty());
}
