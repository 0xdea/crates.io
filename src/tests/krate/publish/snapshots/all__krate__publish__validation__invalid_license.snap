---
source: src/tests/krate/publish/validation.rs
expression: response.into_json()
---
{
  "errors": [
    {
      "detail": "unknown or invalid license expression; see http://opensource.org/licenses for options, and http://spdx.org/licenses/ for their identifiers"
    }
  ]
}
